### DISCUSSION TITLE
Type-narrowing and mixins

### DISCUSSION THREAD

**Discussion by bijij** [microsoft/pyright#2055] [General]
In #2049 I experienced a case where the narrowing scope was seemingly "reset" as pyright was under the impression that code was essentially unreachable, 

I was able to resolve my issue quite easily however I did think of a situation where a resolution might not be as simple.
Consider:

```py
from typing import Optional


class A:
    ...


class MyMixin:
    ...


class B(A):
    ...

class C(A, MyMixin):
    ...

def foo(a: A, b: Optional[int]):
    assert b is not None

    reveal_type(b)  # Pyright:  Type of "b" is "int"

    ...  # do things

    if not isinstance(a, MyMixin):
        ... # do things if the class doesn't inherit from the mixin
        # return at the end
        return

   ...  # do things related to the mixin

    reveal_type(b)  # Pyright:  Type of "b" is "int | None"

```

It's possible there are many classes which inherit from `A` and this function perhaps checks whether a mixin class is in the MRO.
How would one handle a situation like this?

Inherently it's shuffling into Intersection type territory which obviously (and in my opinion unfortunately) python typing does not support.

Other than writing a seperate function which returns the isinstance result but **not** type-hinting it as a typeguard. e.g.

```py
def is_instance(obj: Any, class_or_tuple: Union[Type, Tuple[Type, ...]]) -> bool:
    return isinstance(obj, class_or_tuple)
```

I don't see a practical solution.

Thoughts?

**Comment by erictraut** [comment]
This is related to issue https://github.com/microsoft/pyright/issues/1851, which has been open for a while. I've been thinking about the best way to handle it. I wasn't fully convinced that I should fix it because a "fix" will mask other common bugs. It will effectively render the `reportUnnecessaryIsInstance` diagnostic check useless. However, it would work with mixins and other forms of multiple inheritance.

I think that your example above is pretty compelling, so I'm going to make an attempt to address this. My plan is to synthesize an intersection type — effectively creating a class that subclasses from both the known type(s) and the type(s) specified in the `isinstance` or `issubclass` call.

**Reply by erictraut** [reply]
I've updated the type narrowing logic as indicated above. This addresses the example you provided at the start of this thread. It also addresses the issue reported in #2049.

**Reply by bijij** [reply]
Wonderful! 

A faux-intersection type is a great approach, I have a question about this though:
In a situation where the two intersection types share an attribute of differing type, what does pyright do?

My brain tells me the type of the attribute should be a Union of the two differing types, or the intersection should just be disallowed. but I'm curious nonetheless.

Thank you for all the work you do on this project.

**Reply by erictraut** [reply]
Currently, one attribute type will "win" over the other based on MRO (method resolution order). That's the behavior for multiple inheritance if you were to explicitly declare the class. Of course, if you were declaring the class, you could specify the ordering of the base classes, which would influence the MRO.

In any case, I already consider this an edge case. Overlapping attributes within a synthesized intersection type is an edge case on top of an edge case, so I'm not concerned about it at this point.
